### ISSUE TITLE
RFC: Von Mises distribution (circular) moments

### ISSUE THREAD

**Issue by dschmitz89** [scipy/scipy#19798]
### Describe your issue.

Currently, `vonmises.stats()` is very slow and does not compute the circular variance. While there are different definitions of circular kurtosis from a quick literature scan, the circular variance is pretty much standardized. The formula for the circular variance of the von Mises distribution is available: https://en.wikipedia.org/wiki/Von_Mises_distribution#Moments.

Fixing this would be easy to do (see below) but it breaks backwards compatability. Thinking of the general distribution infrastructure at the moment, I would like to gather feedback if this should be done now or later.

### Reproducing Code Example

```python
from scipy.stats import vonmises, circvar, circmean
from scipy.special import i0e, i1e

loc, kappa = 3, 10

vonmises.stats(loc=loc, kappa=kappa, moments="mvs")  # takes 2 sec on my machine
>>> (3.0, 0.10565504392764057, -1.75350482426206e-15)

samples = vonmises.rvs(loc=loc, kappa=kappa, size=100_000)
circmean(samples), circvar(samples)
>>> (2.9999484119189654, 0.05140615221071776)

# using analytical formulas for the moments
var_analytic = 1 - i1e(kappa)/i0e(kappa)  # takes 1 mu sec
mean = loc
skew = 0  # von mises distribution is symmetric
mean, var_analytic, skew  
>>> (3, 0.05140017404515407, 0)
```

### Error message

```shell
N/A
```

### SciPy/NumPy/Python version and system information

```shell
N/A
```

**Comment by dschmitz89** [comment]
CC @fancidev : your new function from #20457 would come in handy for the circular cariance here. :)

**Comment by fancidev** [comment]
Glad to find its use here! :-) 

I also read from Wikipedia that higher order circular moment is computed as $I_n(x)/I_0(x)$. For small $n$ (like below 5) this is easy to compute by multiplying together the adjacent ratios; for larger $n$ unfortunately I don’t know of a better way than `ive(n,x)/ive(0,x)`.

> I would like to gather feedback if this should be done now or later.

Mentally I think of an $n$-D (directional) distribution as an $n$-D random vector constrained to some (spherical) surface. But then I find it not straightforward to reconcile circular moment with “traditional” moment definitions. Therefore I’d personally prefer to find a proper place for circular statistics in the upcoming new infrastructure rather than “overloading” the existing `stats` method to return circular stats.

**Comment by mdhaber** [comment]
I started implementing a `CircularDistribution` subclass of the new `ContinuousDistribution` a few months ago, but quickly found that there are a lot of interface decisions to be made. It would be helpful if others would research existing interfaces and think very carefully about how we might introduce one that feels like the new continuous distributions but diverges as needed. For instance, is it useful to define a one-argument CDF that assumes an origin point, or does it only really make sense to ask about the cumulative probability between two points? What does that mean for the inverse CDF? I'd like for at least two people to do this besides me so there's less change of exactly two conflicting opinions.

**Comment by dschmitz89** [comment]
> I started implementing a `CircularDistribution` subclass of the new `ContinuousDistribution` a few months ago, but quickly found that there are a lot of interface decisions to be made. It would be helpful if others would research existing interfaces and think very carefully about how we might introduce one that aligns well with the new continuous distributions but diverges as needed. For instance, is it useful to define a one-argument CDF that assumes an origin point, or does it only really make sense to ask about the cumulative probability between two points? What does that mean for the inverse CDF? I'd like for at least two people to do this besides me so there's less change of exactly two conflicting opinions.

I will gladly join the discussion about that :). Is it the same branch as your other distribution infra work?

**Comment by mdhaber** [comment]
Well, yeah, the very little I did is in that branch:
https://github.com/mdhaber/scipy/blob/29543a95ecc3db1dcf10eea8c163957649a71cc8/scipy/stats/_new_distributions.py#L359

But let's keep that issue focused on the basic infrastructure. Maybe generalize this issue to discuss the interface of the new circular distributions (not worrying about implementation yet)? For starters, I think it would be useful to just write out the signatures and short descriptions of the methods. For instance:
A circular distribution is parameterized by the *shape* parameters of the underlying distribution (e.g. von Mises kappa) and parameters `a` and `b` (default `(-np.pi, np.pi)`). Methods include:

- `pdf(x)` - the PDF evaluated at argument `x` wrapped to the interval `a, b`. (What is the value at `x=oo`? NaN?)
- `cdf(x, y)` - the cumulative probability between `x` and `y`. (Is it OK for `y` to be `None`, in which case perhaps this is the cumulative probability between `a, x`? What happens if `y - x > pi`? Is there a difference if `y > x`? Etc...)
- `mean()` - the _circular_ mean
- `var()` - the _circular_ variance. (This would not be scaled in any way by `a`/`b`, right?)
- etc.

**Comment by fancidev** [comment]
I agree with @mdhaber that it’s not straightforward to specialize `ContinuousDistribution` to a `CircularDistribution` because of many inconsistencies in certain definitions. I guess it might be more natural and practical if we don’t subclass `CircularDistribution` from `ContinuousDistribution` and instead make a class of its own. Let me elaborate why. I’ll summarise the concept/definition differences between a “continuous distribution” and a circular distribution, essentially by extracting pieces of information from Chapter 3 of Mardia and Supp (1999). Then I’ll explain why these differences warrants to have a “standalone” `CirculatDistribution` class that is not derived from `ContinuousDistribution.` 

**Reference**

Mardia, K. V. and Jupp, P. E. (1999). _Directional Statistics_. John Wiley & Sons, Inc.

**Comment by fancidev** [comment]
**Differences between circular and regular distribution**

(To be completed)

**Comment by fancidev** [comment]
**Interface suggestion**

(To be completed)

**Comment by mdhaber** [comment]
Again @fancidev , let's focus on the interface, not the implementation.

**Comment by rkern** [comment]
I, for one, find it nearly impossible to reason about an interface without 2 implementations in mind (at least sketches of them). Thinking about (possible) implementations illuminates the constraints.

**Comment by fancidev** [comment]
I have completed a spec for a `CircularDistribution` class; please see the three consecutive comments above.

It could be convenient to read the last comment first, which lays out the interface. To explain why certain methods are defined the way they are, I list in the second comment the basic differences between regular and circular distributions. Finally, an attempt is made in the first comment to rationalize why these differences exist, where I write down my own understanding as a way to kick off the discussion.

I hope this could be helpful!

**Comment by mdhaber** [comment]
Thanks @fancidev. @dschmitz89 it would be terrific if you could write a spec independently, and then we can compare.

**Comment by NeilGirdhar** [comment]
I like nearly everything @fancidev wrote, but:

> Where it exists, the density function of a probability distribution integrates to **one** over the real line.

I would say it integrates to one over the support.

> The second group of methods share name with those of a regular distribution but have extended semantics that are not fully compatible.

With the above definition, you can say that `pdf`, etc. are fully compatible.

This will keep things simple when you also have distributions whose support is the simplex.

I think the one-argument form of `cdf` will probably be more useful since various invariants like numerical calculation of `pdf` in terms of `cdf` are easier to implement this way?

Maybe one day it might help to add `canonicalize_to_support(x)` that maps `x` to `[a, b]` for circular distributions and does nothing for ordinary supports?  This might also be useful for distributions on the simplex to map `x` to some vector whose elements sum to one (projection onto the simplex)?

I guess `a` and `b` can be recovered by doing `icdf(0)` and `icdf(1)`?
